library/test/src/formatters/junit.rs

```diff
@@ -132,3 +136,39 @@ impl<T: Write> OutputFormatter for JunitFormatter<T> {
         Ok(state.failed == 0)
     }
 }
+
+fn parse_class_name(desc: &TestDesc) -> (String, String) {
+    match desc.test_type {
+        TestType::UnitTest => parse_class_name_unit(desc),
+        TestType::DocTest => parse_class_name_doc(desc),
+        TestType::IntegrationTest => parse_class_name_integration(desc),
+        TestType::Unknown => (String::from("unknown"), String::from(desc.name.as_slice())),
+    }
+}
+
+fn parse_class_name_unit(desc: &TestDesc) -> (String, String) {
+    // Module path => classname
+    // Function name => name
+    let module_segments: Vec<&str> = desc.name.as_slice().split("::").collect();
+    let (class_name, test_name) = match module_segments[..] {
+        [test] => (String::from("crate"), String::from(test)),
+        [ref path @ .., test] => (path.join("::"), String::from(test)),
```

Comment:
Yes. I thought about it and figured it's an edge case to have tests outside the designated  module, so might not worth it?

When I made that decision I didn't have a separate way to handle unit/integration/doc tests, though, so I didn't like the idea of prepending `crate::` .

Reply:
Thinking about it a bit I think you made the right call. If you imagine these as actual paths that are referring to concrete items in a crate then `crate` here is equivalent to `self`, and if this were actual rust code then that alias would presumably be valid in this context, since `self` is the root module of the crate.